bundles/org.eclipse.search.core/search/org/eclipse/search/internal/core/text/TextSearchVisitor.java

```diff
@@ -472,8 +472,9 @@ private boolean hasBinaryContent(CharSequence seq, IFile file) throws CoreExcept
 	private List<TextSearchMatchAccess> locateMatches(IFile file, CharSequence searchInput, Matcher matcher, IProgressMonitor monitor) throws CoreException {
 		List<TextSearchMatchAccess> occurences= null;
 		matcher.reset(searchInput);
-		int k= 0;
-		while (matcher.find()) {
+		// Check for cancellation before calling matcher.find() since that call
+		// can be very expensive
+		while (!monitor.isCanceled() && matcher.find()) {
```

Comment:
I wonder why the check was explicitly done not at every match. Could it be, monitor check is too expensive (not sure which monitor is used here and if that is a wrapper to some UI progress monitor that does more than just boolean check)?

Reply:
I do recall cancel checking to have some significant overhead. Has anyone debug stepped through the cancel check to inspect manually what the overhead might be?

Reply:
Stepping through it shows that this is a submonitor and it goes all the way to `JobInfo`...

```java
Thread [Worker-4: File Search Worker] (Suspended)	
	JobInfo.isCanceled() line: 315	
	ProgressManager$JobMonitor.isCanceled() line: 286	
	SubMonitor$RootInfo.isCanceled() line: 278	
	SubMonitor.isCanceled() line: 661	
	TextSearchVisitor.locateMatches(IFile, CharSequence, Matcher, IProgressMonitor) line: 477	
	TextSearchVisitor$TextSearchJob.processFile(List<IFile>, IProgressMonitor) line: 212	
	TextSearchVisitor$TextSearchJob.run(IProgressMonitor) line: 173	
	Worker.run() line: 63	
```

But the calls are not expensive. I ran the search specified in #1801 (an expensive regex) and the sampling says that checking for cancelation doesn't really need much time. 

I sampled with 10ms. These are the "hotspots" for `isCanceling`:
![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/2205684/f3fc0283-c5aa-496d-ae9a-6a82f4931c0f)

And these for `locateMatches`:
![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/2205684/7d53488f-5539-461a-b41a-2b291ec18398)

Here's the sample: 
[Search_Regex_10ms_sampling.zip](https://github.com/eclipse-platform/eclipse.platform.ui/files/14917332/Search_Regex_10ms_sampling.zip)

The part of the code that checks only after 20 iterations was added [over 20 years ago](https://github.com/eclipse-platform/eclipse.platform.ui/commit/b970afafe42296adfbbcdd386ffbee1490ac28ab#diff-3fee34fc300739c9542e87b8a8a8183655c798385c8a920508f0012db42f4947R118) (see b970afafe42296adfbbcdd386ffbee1490ac28ab). Maybe it was expensive back then?

Reply:
Time changes all things. Thanks for checking!